xray/plot/facetgrid.py

```diff
@@ -0,0 +1,307 @@
+        self._row_var = row
+        self._col_var = col
+        self._col_wrap = col_wrap
+
+        self.set_titles()
+
+    def __iter__(self):
+        return self.axes.flat
+
+    def map_dataarray(self, plotfunc, *args, **kwargs):
+        """Apply a plotting function to a 2d facet's subset of the data.
+
+        This is more convenient and less general than the map method.
+
+        Parameters
+        ----------
+        plotfunc : callable
+            A plotting function with the same signature as a 2d xray
+            plotting method such as `xray.plot.imshow`
+        args :
+            positional arguments to plotfunc
+        kwargs :
+            keyword arguments to plotfunc
+
+        Returns
+        -------
+        self : FacetGrid object
+            the same FacetGrid on which the method was called
+
+        """
+        import matplotlib.pyplot as plt
+
+        # These should be consistent with xray.plot._plot2d
+        cmap_kwargs = {'plot_data': self.darray.values,
+                'vmin': None,
+                'vmax': None,
+                'cmap': None,
+                'center': None,
+                'robust': False,
+                'extend': None,
+                'levels': 7 if 'contour' in plotfunc.__name__ else None, # MPL default
+                'filled': plotfunc.__name__ != 'contour',
+                }
+
+        # Allow kwargs to override these defaults
+        for param in kwargs:
+            if param in cmap_kwargs:
+                cmap_kwargs[param] = kwargs[param]
+
+        # colormap inference has to happen here since all the data in self.darray
+        # is required to make the right choice
+        cmap_params = _determine_cmap_params(**cmap_kwargs)
+
+        if 'contour' in plotfunc.__name__:
+            # extend is a keyword argument only for contour and contourf, but
+            # passing it to the colorbar is sufficient for imshow and
+            # pcolormesh
+            kwargs['extend'] = cmap_params['extend']
+            kwargs['levels'] = cmap_params['levels']
+
+        defaults = {
+                'add_colorbar': False,
+                'add_labels': False,
+                'norm': cmap_params.pop('cnorm'),
+                }
+
+        # Order is important
+        defaults.update(cmap_params)
+        defaults.update(kwargs)
+
+        for d, ax in zip(self.name_dicts.flat, self.axes.flat):
+            # Handle the sentinel value
+            if d is not None:
+                subset = self.darray.loc[d]
+                plotfunc(subset, ax=ax, *args, **defaults)
+
+        # Plot the last subset again to determine x and y values
+        try:
+            dummyfig = plt.figure()
+            dummyax = dummyfig.add_axes((0, 0, 0, 0))
+            defaults['add_labels'] = True
+
+            mappable = plotfunc(subset,
+                    ax=dummyax, *args, **defaults)
+
+            xlab, ylab = dummyax.get_xlabel(), dummyax.get_ylabel()
```

Comment:
Instead I'm changing the signature to 
```
def map_dataarray(self, plotfunc, x, y, *args, **kwargs):
```
which is probably more clear for the reader of the docs